pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -169,56 +169,76 @@ async def consumer(
+                    message_queue.task_done()
+                    semaphore.release()
+                    continue
 
-            metrics_client.histogram(
-                'pubsub.consumer.latency.receive',
-                # publish_time is in UTC Zulu
-                # https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage
-                time.time() - message.publish_time.timestamp())
-
-            task = asyncio.ensure_future(_execute_callback(
-                message,
-                callback,
-                ack_queue,
-                nack_queue,
-                metrics_client,
-            ))
-            task.add_done_callback(lambda _f: semaphore.release())
-            message_queue.task_done()
+                metrics_client.histogram(
+                    'pubsub.consumer.latency.receive',
+                    # publish_time is in UTC Zulu
+                    # https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage
+                    time.time() - message.publish_time.timestamp())
+
+                task = asyncio.ensure_future(_execute_callback(
+                    message,
+                    callback,
+                    ack_queue,
+                    nack_queue,
+                    metrics_client,
+                ))
+                task.add_done_callback(lambda _f: semaphore.release())
+                message_queue.task_done()
+        except asyncio.CancelledError:
+            log.info('Consumer worker cancelled. Gracefully terminating...')
+            for _ in range(max_tasks):
+                await semaphore.acquire()
+            await ack_queue.join()
+            if nack_queue:
+                await nack_queue.join()
+            log.info('Consumer terminated gracefully.')
+            raise
 
     async def producer(
             subscription: str,
             message_queue: MessageQueue,
             subscriber_client: 'SubscriberClient',
             max_messages: int,
             metrics_client: MetricsAgent) -> None:
-        while True:
-            try:
-                new_messages = await subscriber_client.pull(
-                    subscription=subscription, max_messages=max_messages)
-            except (asyncio.TimeoutError, KeyError):
-                continue
+        try:
+            while True:
+                try:
+                    new_messages = await subscriber_client.pull(
+                        subscription=subscription, max_messages=max_messages)
+                except (asyncio.TimeoutError, KeyError):
+                    continue
+
+                pulled_at = time.perf_counter()
+                for m in new_messages:
+                    await message_queue.put((m, pulled_at))
 
+                metrics_client.histogram(
+                    'pubsub.producer.batch', len(new_messages))
+
+                await message_queue.join()
+        except asyncio.CancelledError:
```

Comment:
The behavior here looks like it will depend on when and where we get cancelled in lines 217-233 above. If we get cancelled before we even make one successful pull, I don't think `new_messages` will be defined and will result in a `NameError`. This will be caught and handled elsewhere, though.

If we get cancelled on https://github.com/talkiq/gcloud-aio/blob/am/pubsub/graceful_shutdown/pubsub/gcloud/aio/pubsub/subscriber.py#L220 after at least one pull, then I think the `new_messages` we end up putting in the queue on https://github.com/talkiq/gcloud-aio/blob/am/pubsub/graceful_shutdown/pubsub/gcloud/aio/pubsub/subscriber.py#L237 will be those from the previous iteration (so processing them twice)

Same goes for elsewhere in the above code block.. some messages could be processed twice. Any issues with this?

Perhaps we can initialize each iteration with `new_messages=[]`, and then change the for loop to something like 

```python
                while new_messages:
                    await message_queue.put((new_messages[-1], pulled_at))
                    new_messages.pop()
```

(I wanted to pop() on the same line as the `await` but that could result in a dropped message)

Reply:
oh yes, very good point! 🙏